File: typescript/packages/sdk/src/lib/assemble.ts
import type { TransactionDetails } from '../types';

/**
 * Join an array of (possibly undefined) strings with a pipe, replacing undefined with empty string.
 */
function joinPipe(fields: Array<string | undefined>): string {
  return fields.map((f) => f ?? '').join('|');
}

/**
 * Build the sectioned merchant_request string for WEB/MOBILE flows.
 */
export function buildMerchantRequestPlain(details: TransactionDetails): string {
  const txn_details = joinPipe([
    details.aggregatorId,
    details.merchantId,
    details.orderNumber,
    details.amount,
    details.country,
    details.currency,
    details.transactionType,
    details.successUrl ?? '',
    details.failureUrl ?? '',
    details.channel,
  ]);

  const pg_details = joinPipe([
    details.pgId,
    details.paymode,
    details.schemeId,
    details.walletType,
  ]);
  const card_details = joinPipe([
    details.cardNumber,
    details.expiryMonth,
    details.expiryYear,
    details.cvv,
    details.cardName,
  ]);

  const cust_details = joinPipe([
    details.customerName ?? '',
    details.customerEmail ?? '',
    details.customerMobile ?? '',
    details.uniqueId ?? '',
    details.isLoggedIn ?? 'Y',
  ]);

  const bill_details = joinPipe([
    details.billAddress,
    details.billCity,
    details.billState,
    details.billCountry,
    details.billZip,
  ]);
  const ship_details = joinPipe([
    details.shipAddress,
    details.shipCity,
    details.shipState,
    details.shipCountry,
    details.shipZip,
    details.shipDays,
    details.addressCount,
  ]);
  const item_details = joinPipe([
    details.itemCount,
    details.itemValue,
    details.itemCategory,
  ]);
  // Section 8 in web flow is reserved empty; API expects fraud_details here (kept empty unless specified)
  const fraud_details = '';
  // Section 9: udf_details (web: 5 fields), API other_details may accept 7
  const udf_details = details.channel === 'API'
    ? joinPipe([
      details.udf1,
      details.udf2,
      details.udf3,
      details.udf4,
      details.udf5,
      details.udf6,
      details.udf7,
    ])
    : joinPipe([
      details.udf1,
      details.udf2,
      details.udf3,
      details.udf4,
      details.udf5,
    ]);

  return [
    txn_details,
    pg_details,
    card_details,
    cust_details,
    bill_details,
    ship_details,
    item_details,
    fraud_details,
    udf_details,
  ].join('~');
}

/**
 * Build the API merchantRequest JSON string with exact field names per spec.
 */
export function buildApiMerchantRequestPlain(details: TransactionDetails): string {
  const apiPayload = {
    card_details: {
      cardNumber: details.cardNumber ?? '',
      expiryMonth: details.expiryMonth ?? '',
      expiryYear: details.expiryYear ?? '',
      cvv: details.cvv ?? '',
      cardName: details.cardName ?? '',
    },
    other_details: {
      udf1: details.udf1 ?? '',
      udf2: details.udf2 ?? '',
      udf3: details.udf3 ?? '',
      udf4: details.udf4 ?? '',
      udf5: details.udf5 ?? '',
      udf6: details.udf6 ?? '',
      udf7: details.udf7 ?? '',
    },
    ship_details: {
      shipAddress: details.shipAddress ?? '',
      shipCity: details.shipCity ?? '',
      shipState: details.shipState ?? '',
      shipCountry: details.shipCountry ?? '',
      shipZip: details.shipZip ?? '',
      shipDays: details.shipDays ?? '',
      addressCount: details.addressCount ?? '',
    },
    txn_details: {
      agId: details.aggregatorId,
      meId: details.merchantId,
      orderNo: details.orderNumber,
      amount: details.amount,
      country: details.country,
      currency: details.currency,
      transactionType: details.transactionType,
      sucessUrl: details.successUrl ?? '',
      failureUrl: details.failureUrl ?? '',
      channel: 'API',
    },
    item_details: {
      itemCount: details.itemCount ?? '',
      itemValue: details.itemValue ?? '',
      itemCategory: details.itemCategory ?? '',
    },
    cust_details: {
      customerName: details.customerName ?? '',
      emailId: details.customerEmail ?? '',
      mobileNumber: details.customerMobile ?? '',
      uniqueId: details.uniqueId ?? '',
      isLoggedIn: details.isLoggedIn ?? 'Y',
    },
    pg_details: {
      pg_Id: details.pgId ?? '',
      paymode: details.paymode ?? '',
      scheme_Id: details.schemeId ?? '',
      wallet_type: details.walletType ?? '',
    },
    bill_details: {
      billAddress: details.billAddress ?? '',
      billCity: details.billCity ?? '',
      billState: details.billState ?? '',
      billCountry: details.billCountry ?? '',
      billZip: details.billZip ?? '',
    },
  };
  return JSON.stringify(apiPayload);
}

File: dart/sdk/lib/src/assemble.dart
import 'dart:convert';
import 'types.dart';

String _joinPipe(List<String?> fields) => fields.map((f) => f ?? '').join('|');

/// Build the sectioned merchant_request string for WEB/MOBILE flows.
String buildMerchantRequestPlain(TransactionDetails d) {
  final txnDetails = _joinPipe([
    d.aggregatorId,
    d.merchantId,
    d.orderNumber,
    d.amount,
    d.country,
    d.currency,
    d.transactionType,
    d.successUrl,
    d.failureUrl,
    d.channel,
  ]);

  final pgDetails = _joinPipe([d.pgId, d.paymode, d.schemeId, d.walletType]);
  final cardDetails = _joinPipe([d.cardNumber, d.expiryMonth, d.expiryYear, d.cvv, d.cardName]);
  final custDetails = _joinPipe([d.customerName, d.customerEmail, d.customerMobile, d.uniqueId, d.isLoggedIn ?? 'Y']);
  final billDetails = _joinPipe([d.billAddress, d.billCity, d.billState, d.billCountry, d.billZip]);
  final shipDetails =
      _joinPipe([d.shipAddress, d.shipCity, d.shipState, d.shipCountry, d.shipZip, d.shipDays, d.addressCount]);
  final itemDetails = _joinPipe([d.itemCount, d.itemValue, d.itemCategory]);
  final reserved = '';
  final udfDetails = _joinPipe([d.udf1, d.udf2, d.udf3, d.udf4, d.udf5]);

  return [
    txnDetails,
    pgDetails,
    cardDetails,
    custDetails,
    billDetails,
    shipDetails,
    itemDetails,
    reserved,
    udfDetails,
  ].join('~');
}

/// Build the API JSON string with exact field names per spec.
String buildApiMerchantRequestPlain(TransactionDetails d) {
  final payload = {
    'card_details': {
      'cardNumber': d.cardNumber ?? '',
      'expiryMonth': d.expiryMonth ?? '',
      'expiryYear': d.expiryYear ?? '',
      'cvv': d.cvv ?? '',
      'cardName': d.cardName ?? '',
    },
    'other_details': {
      'udf1': d.udf1 ?? '',
      'udf2': d.udf2 ?? '',
      'udf3': d.udf3 ?? '',
      'udf4': d.udf4 ?? '',
      'udf5': d.udf5 ?? '',
      'udf6': d.udf6 ?? '',
      'udf7': d.udf7 ?? '',
    },
    'ship_details': {
      'shipAddress': d.shipAddress ?? '',
      'shipCity': d.shipCity ?? '',
      'shipState': d.shipState ?? '',
      'shipCountry': d.shipCountry ?? '',
      'shipZip': d.shipZip ?? '',
      'shipDays': d.shipDays ?? '',
      'addressCount': d.addressCount ?? '',
    },
    'txn_details': {
      'agId': d.aggregatorId,
      'meId': d.merchantId,
      'orderNo': d.orderNumber,
      'amount': d.amount,
      'country': d.country,
      'currency': d.currency,
      'transactionType': d.transactionType,
      'sucessUrl': d.successUrl,
      'failureUrl': d.failureUrl,
      'channel': 'API',
    },
    'item_details': {
      'itemCount': d.itemCount ?? '',
      'itemValue': d.itemValue ?? '',
      'itemCategory': d.itemCategory ?? '',
    },
    'cust_details': {
      'customerName': d.customerName ?? '',
      'emailId': d.customerEmail ?? '',
      'mobileNumber': d.customerMobile ?? '',
      'uniqueId': d.uniqueId ?? '',
      'isLoggedIn': d.isLoggedIn ?? 'Y',
    },
    'pg_details': {
      'pg_Id': d.pgId ?? '',
      'paymode': d.paymode ?? '',
      'scheme_Id': d.schemeId ?? '',
      'wallet_type': d.walletType ?? '',
    },
    'bill_details': {
      'billAddress': d.billAddress ?? '',
      'billCity': d.billCity ?? '',
      'billState': d.billState ?? '',
      'billCountry': d.billCountry ?? '',
      'billZip': d.billZip ?? '',
    },
  };
  return jsonEncode(payload);
}
